gpii/node_modules/flowManager/src/PCPChannel.js

```diff
@@ -0,0 +1,207 @@
+                // avoid notifying the client for the init of LifecycleManagerSession, simplifies test and clients
+                excludeSource: "init"
+            }
+        }
+    },
+    modelListeners: {
+        updatePreferences: {
+            path: "{flowManager}.pcpChannel.model.preferences",
+            includeSource: "PCP",
+            funcName: "gpii.pcpChannel.updatePreferences",
+            args: ["{flowManager}.lifecycleManager", "{change}.value"]
+        },
+        updateContext: {
+            path: "{flowManager}.pcpChannel.model.activeContextName",
+            includeSource: "PCP",
+            funcName: "gpii.pcpChannel.updateContext",
+            args: ["{flowManager}.contextManager", "{pcpChannel}", "{change}.value"]
+        }
+    }
+});
+
+fluid.defaults("gpii.pcpChannel", {
+    gradeNames: ["fluid.modelComponent"],
+    members: {
+        outputBlocked: null
+    },
+    distributeOptions: {
+        // Not {lifecycleManager session} because of FLUID-6193
+        target: "{flowManager lifecycleManager session}.options.gradeNames",
+        record: "gpii.pcpChannel.sessionBinder"
+    },
+    listeners: {
+        "{lifecycleManager}.events.onSessionStop": {
+            namespace: "pcpChannel",
+            funcName: "gpii.pcpChannel.sessionStop",
+            args: "{pcpChannel}"
+        }
+    }
+});
+
+fluid.logObjectRenderChars = 10240;
+
+gpii.pcpChannel.updatePreferences = function (lifecycleManager, newPrefs) {
+    // TODO: to be implemented - must run a "mini-matchmaker cycle" via the lifecycleManager or otherwise
+    fluid.log("Received preferences update from PCP UI of ", newPrefs);
+};
+
+gpii.pcpChannel.updateContext = function (contextManager, pcpChannel, newContext) {
+    fluid.log("Received context update from PCP UI of ", newContext);
+    // Abominable hack to avoid confusing client with numerous update messages. The ChangeApplier should really support
+    // "manifest transactions" / "vertical transactions"
+    pcpChannel.outputBlocked = fluid.promise();
+    var clearBlock = function () {
+        delete pcpChannel.outputBlocked;
+    };
+    pcpChannel.outputBlocked.then(clearBlock, clearBlock);
+    pcpChannel.applier.change([], null, "DELETE");
+    var promise = contextManager.contextChanged(newContext);
+    if (promise) {
+        fluid.promise.follow(promise, pcpChannel.outputBlocked);
+    } else {
+        pcpChannel.outputBlocked.resolve();
+    }
+};
+
+gpii.pcpChannel.filterSolution = function (solution) {
+    return {
+        name: solution.name,
+        settingsHandlers: fluid.transform(solution.settingsHandlers, function (oneHandler) {
+            return fluid.filterKeys(oneHandler, ["supportedSettings"]);
+        })
+    };
+};
+
+gpii.pcpChannel.filterSchema = function (schema) {
+    return schema && fluid.censorKeys(schema, "default");
+};
+
+gpii.pcpChannel.emitSettingControl = function (settingControls, schema, prefVal, keySegs, solutionName) {
```

Comment:
comments on this function - in particular that settingsControls will be modified by it